clippy_lints/src/derivable_impls.rs

```diff
@@ -0,0 +1,114 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::{in_macro, is_automatically_derived, is_default, match_def_path, paths};
+use rustc_hir::{Block, Expr, ExprField, ExprKind, Impl, ImplItemKind, Item, ItemKind, Node, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for double comparisons that could be simplified to a single expression.
+    ///
+    ///
+    /// ### Why is this bad?
+    /// Readability.
+    ///
+    /// ### Example
+    /// ```rust
+    /// # let x = 1;
+    /// # let y = 2;
+    /// if x == y || x < y {}
+    /// ```
+    ///
+    /// Could be written as:
+    ///
+    /// ```rust
+    /// # let x = 1;
+    /// # let y = 2;
+    /// if x <= y {}
+    /// ```
+    pub DERIVABLE_IMPLS,
+    complexity,
+    "manual implementation of a trait which is equal to a derive"
+}
+
+declare_lint_pass!(DerivableImpls => [DERIVABLE_IMPLS]);
+
+fn struct_fields<'hir>(e: &'hir Expr<'hir>) -> Option<&'hir [ExprField<'hir>]> {
+    match e.kind {
+        ExprKind::Struct(_, fields, _) => Some(fields),
+        ExprKind::Block(Block { expr, .. }, _) => expr.and_then(struct_fields),
+        _ => None,
+    }
+}
+
+fn is_path_self(e: &Expr<'_>) -> bool {
+    if let ExprKind::Path(QPath::Resolved(_, p)) = e.kind {
+        if p.segments.len() != 1 {
+            return false;
+        }
+        if p.segments[0].ident.name == rustc_span::symbol::kw::SelfUpper {
+            return true;
+        }
```

Comment:
You can do `matches!(path.res, Res::SelfTy(..))`

Reply:
It doesn't pass tests